File: ExecutionContext/MultipleOrderedEC-C++/mainwindow.cpp
/*!
* @file  mainwindow.cpp
* @brief メインウインドウ
*
*/

#include <QLabel>
#include <QtWidgets>

#include "mainwindow.h"
#include "FileStreamFunc.h"






/**
*@brief メインウインドウのコンストラクタ
* @param ec 実行コンテキストオブジェクト
*/
MainWindow::MainWindow(RTC::MultipleOrderedEC *ec)
{
	
	
	tc = QTextCodec::codecForLocale();

	
	
    QVBoxLayout *layout = new QVBoxLayout;

	 
	

	this->m_ec = ec;
	SC = new SetComp(ec);
	connect(SC, SIGNAL(UpdateSizeSignal(int , int )),
            this, SLOT(m_resize(int , int )));

	layout->addWidget(SC);

	layout->addStretch();
	UB = new QPushButton(tc->toUnicode("更新"));
	layout->addWidget(UB);

	connect(UB, SIGNAL(clicked()),
            this, SLOT(UpdateComp()));

	QPushButton *DB = new QPushButton(tc->toUnicode("追加"));
	layout->addWidget(DB);

	connect(DB, SIGNAL(clicked()),
            SC, SLOT(CreateComp()));

	

    widget = new QWidget;
    widget->setLayout(layout);

	
	QScrollArea *area = new QScrollArea();
	area->setWidget(widget);
	
	

    setCentralWidget(area);
    setWindowTitle(tr("MultipleOrderedECGUI"));
    setUnifiedTitleAndToolBarOnMac(true);

	
	SC->UpdateSizeSlot();


	createAction();
	createMenus();
}

/**
*@brief サイズを変更するときに呼び出されるスロット
* @param w 幅
* @param h 高さ
*/
void MainWindow::m_resize(int w, int h)
{
	widget->resize(w, h);
}

/**
*@brief RTCが追加、削除されたときに呼び出されるスロット
*/
void MainWindow::UpdateComp()
{
	SC->UpdateComps();
	SC->UpdateComp2();
}

/**
*@brief アクションの作成の関数
*/
void MainWindow::createAction()
{
	newAct = new QAction(tr("&New..."),this);
	newAct->setShortcuts(QKeySequence::New);
    connect(newAct, SIGNAL(triggered()), this, SLOT(newFile()));


	openAct = new QAction(tr("&Open..."),this);
    openAct->setShortcuts(QKeySequence::Open);
    connect(openAct, SIGNAL(triggered()), this, SLOT(open()));


    saveAct = new QAction(tr("&Save"),this);
    saveAct->setShortcuts(QKeySequence::Save);
    connect(saveAct, SIGNAL(triggered()), this, SLOT(save()));

	saveAsAct = new QAction(tr("&Save &As"), this);
	saveAsAct->setShortcuts(QKeySequence::SaveAs);
	connect(saveAsAct, SIGNAL(triggered()), this, SLOT(saveAs()));
}

/**
*@brief メニューの作成の関数
*/
void MainWindow::createMenus()
{
	fileMenu = menuBar()->addMenu(tr("&File"));
	fileMenu->addAction(newAct);
    fileMenu->addAction(openAct);
    fileMenu->addAction(saveAct);
	fileMenu->addAction(saveAsAct);
	

}

/**
*@brief ファイル読み込みスロット
*/
void MainWindow::open()
{
	QString fileName = QFileDialog::getOpenFileName(this,
							tc->toUnicode("開く"), "",
							tr("Config File (*.conf);;Lua File (*.lua);;All Files (*)"));
	

	QByteArray ba = fileName.toLocal8Bit();
	SC->open(ba);

	m_ec->FileName = ba;

}

/**
*@brief ファイル保存のスロット
* @return 成功でTrue、失敗でFalse
*/
bool MainWindow::save()
{
	
	
	if (m_ec->FileName == "")
	{
		
		return saveAs();
	}
	else
	{
		PathList pl = PathList(m_ec->FileName);
		if (pl.cxt == "lua")
		{
			return saveAs();
		}
		else
		{
			return SC->save(m_ec->FileName.c_str());
		}
		
		
	}
}

/**
*@brief 別のファイル保存のスロット
* @return 成功でTrue、失敗でFalse
*/
bool MainWindow::saveAs()
{
	QString fileName = QFileDialog::getSaveFileName(this,
							tc->toUnicode("保存"), "",
							tr("Config File (*.conf);;All Files (*)"));
	if (fileName.isEmpty())
        return false;

	QByteArray ba = fileName.toLocal8Bit();
	

	m_ec->FileName = ba;
    return SC->save(ba);
}

/**
*@brief 初期化のスロット
*/
void MainWindow::newFile()
{
	SC->newFile();

	m_ec->FileName = "";
    
}

/**
*@brief 実行順序をGUIに反映させる関数
* @param rs 並列ブロックのリスト
*/
void MainWindow::UpdateRTC(std::vector<Rule> &rs)
{
	SC->UpdateRTC(rs);
	
}

File: ExecutionContext/MultipleOrderedEC-Python/MultipleOrderedEC_Lib/MainWindow.py
# -*- coding: utf-8 -*-
##
#   @file MainWindow.py
#   @brief メインウィンドウ

import os
from PyQt4 import QtCore, QtGui
from SetComp import SetComp



##
# @class MainWindow
# @brief メインウィンドウ
#
class MainWindow(QtGui.QMainWindow):
    ##
    # @brief コンストラクタ
    # @param self 
    # @param ec 実行コンテキストオブジェクト
    def __init__(self, ec):
        super(MainWindow, self).__init__()

        self.layout = QtGui.QVBoxLayout()
        self.m_ec = ec
        self.SC = SetComp(self.m_ec)
        
        self.SC.UpdateSizeSignal.connect(self.m_resize)
        self.layout.addWidget(self.SC)

        self.layout.addStretch()
        self.UB = QtGui.QPushButton(u"更新")
        self.layout.addWidget(self.UB)


        self.UB.clicked.connect(self.UpdateComp)

        self.DB = QtGui.QPushButton(u"追加")
        self.layout.addWidget(self.DB)

        self.DB.clicked.connect(self.SC.CreateComp)

	
	

        self.widget = QtGui.QWidget()
        self.widget.setLayout(self.layout)
        self.area = QtGui.QScrollArea()
        self.area.setWidget(self.widget)
        self.setCentralWidget(self.area)
        self.setWindowTitle("MultipleOrderedECGUI")
        self.setUnifiedTitleAndToolBarOnMac(True)

        self.SC.UpdateSizeSlot()

        self.newAct = None
        self.openAct = None
        self.saveAct = None
        self.fileMenu = None

        self.createAction()
        self.createMenus()

        #self.widget.resize(400, 400)

    
    ##
    # @brief サイズを変更するときに呼び出されるスロット
    # @param self 
    # @param w 幅
    # @param h 高さ
    def m_resize(self, w, h):

        self.widget.resize(w, h)

    
    ##
    # @brief RTCが追加、削除されたときに呼び出されるスロット
    # @param self 
    def UpdateComp(self):

        self.SC.UpdateComps()
        self.SC.UpdateComp2()

    
    ##
    # @brief アクションの作成の関数
    # @param self 
    def createAction(self):

        self.newAct = QtGui.QAction("&New...",self)
        self.newAct.setShortcuts(QtGui.QKeySequence.New)
        self.newAct.triggered.connect(self.newFile)
        


        self.openAct = QtGui.QAction("&Open...",self)
        self.openAct.setShortcuts(QtGui.QKeySequence.Open)
        self.openAct.triggered.connect(self.open)


        self.saveAct = QtGui.QAction("&Save",self)
        self.saveAct.setShortcuts(QtGui.QKeySequence.Save)
        self.saveAct.triggered.connect(self.save)

        self.saveAsAct = QtGui.QAction("&Save &As",self)
        self.saveAsAct.setShortcuts(QtGui.QKeySequence.SaveAs)
        self.saveAsAct.triggered.connect(self.saveAs)
        

    
    ##
    # @brief メニューの作成の関数
    # @param self 
    def createMenus(self):

        self.fileMenu = self.menuBar().addMenu("&File")
        self.fileMenu.addAction(self.newAct)
        self.fileMenu.addAction(self.openAct)
        self.fileMenu.addAction(self.saveAct)
        self.fileMenu.addAction(self.saveAsAct)




    
    ##
    # @brief ファイル読み込みスロット
    # @param self 
    def open(self):

        fileName = QtGui.QFileDialog.getOpenFileName(self,u"開く","","Config File (*.conf);;Python File (*.py);;All Files (*)")
        

        ba = str(fileName.toLocal8Bit())
        self.SC.open(ba)
        self.m_ec.FileName = ba

    ##
    # @brief ファイル保存スロット
    # @param self
    # @return 成功でTrue、失敗でFalse
    def save(self):
        root, ext = os.path.splitext(self.m_ec.FileName)
        if self.m_ec.FileName == "" or ext == ".py":
            return self.saveAs()
        else:
            return self.SC.save(self.m_ec.FileName)
            
            

    
    ##
    # @brief 別のファイル保存のスロット
    # @param self
    # @return 成功でTrue、失敗でFalse
    def saveAs(self):

        fileName = QtGui.QFileDialog.getSaveFileName(self,u"保存", "","Config File (*.conf);;All Files (*)")
        if fileName.isEmpty():
            return False

        ba = str(fileName.toLocal8Bit())
        self.m_ec.FileName = ba


        return self.SC.save(ba)


    
    ##
    # @brief 初期化のスロット
    # @param self 
    def newFile(self):

        self.SC.newFile()
        self.m_ec.FileName = ""

    


    
    ##
    # @brief 実行順序をGUIに反映させる関数
    # @param self
    # @param rs 並列ブロックのリスト
    def UpdateRTC(self,rs):

        self.SC.UpdateRTC(rs)

